Add docstring example best practices to doc (#5)
pyansys/pyansys-sphinx-theme: doc/source/coding_style.rst

"first-class" users.  Note that due to the complexity of maintaining
multiple "release branches" in a repository, the number of active
release branches should be between one and three.

> Docstring Examples Best-Practice
> ---------------------------------

Reviewer: ```suggestion
Docstring Examples Best-Practice
--------------------------------
```

Defining docstring examples for methods and classes are extremely 
useful. The examples give users an easy place to start when trying 
out the API, by showing them exactly how to operate on a method or 
class. By using ``doctest`` through ``pytest``, docstring examples can 
also be used to perform regression testing to verify that the code is 

> executing as expected.

Reviewer: ```suggestion
executing as expected.

This is an important feature of maintainable documentation as examples
must always match the API they are documenting, and any changes within
the API without a corresponding change in the documentation will
trigger doctest failures.
```

Setting Up ``doctest``
~~~~~~~~~~~~~~~~~~~~~~
First, install ``pytest``:

.. code::

    pip install pytest

Then, ``doctest`` can be run on any Python file by running:

.. code::

    pytest --doctest-modules file.py

Here, ``doctest`` will search for any examples in the docstrings and will then
execute them to verify that they function as written.

Using ``pytest`` Fixtures
~~~~~~~~~~~~~~~~~~~~~~~~~

> To define a set up sequence before the ``doctest`` run or before a given 

Reviewer: ```suggestion
To define a setup sequence before the ``doctest`` run or before a given 
```
I think "setup" makes more sense here as a noun vs "set up" which is two words as a verb.

Author: Yes, you're right. I just keep writing setup since I am so used to the AEDT concept "Analysis Setup"!

module is tested, ``pytest`` fixtures can be used. Fixtures allow docstring 

> examples to access shared objects, so there is no need to repeat the set 
> up in each example.

Reviewer: ```suggestion
examples to access shared objects, so there is no need to repeat the setup
in each example.
```
I think using "setup" as a noun makes more sense here, but feel free to ignore this.

``Icepak`` object as the value. 

.. code:: python

    import pytest
    from pyaedt import Icepak

    @pytest.fixture(autouse=True, scope="module")
    def create_icepak(doctest_namespace):
        doctest_namespace["icepak"] = Icepak(projectname="Project1", designname="IcepakDesign1")

Then, the ``Icepak`` object can be used directly inside a docstring
example, by referencing the key ``icepak``.

.. code:: python

    def assign_openings(self, air_faces):
        """Assign openings to a list of faces.

        Parameters
        ----------
        air_faces : list
            List of face names.

        Returns
        -------
        :class:`pyaedt.modules.Boundary.BoundaryObject`
            Boundary object when successful or ``None`` when failed.

        Examples
        --------

        Create an opening boundary for the faces of the "USB_GND" object.

        >>> faces = icepak.modeler.primitives["USB_GND"].faces
        >>> face_names = [face.id for face in faces]
        >>> boundary = icepak.assign_openings(face_names)
        pyaedt Info: Face List boundary_faces created
        pyaedt Info: Opening Assigned

        """

Useful Features

> ---------------

Reviewer: Now that I think of it, we probably should create a directory with an `index.rst` and include different sections regarding ansys specific coding style and notes.  This document isn't too bad at the moment, but when kathy starts contributing, I think we'll have more to add.

We don't have to implement this now, but eventually we should have:

```
coding_style/
    index.rst
    coding_style.rst
    best_practices.rst
    docstring_examples.rst
```

where `docstring_examples.rst` contains..

```
Docstring Examples Best-Practices
=================================

General Notes
-------------------
Add general notes on how to write a docstring example.

Documentation Testing
---------------------

Setting Up ``doctest``
~~~~~~~~~~~~~~~~~~~~~~

Using ``pytest`` Fixtures
~~~~~~~~~~~~~~~~~~~~~~~~~

Using Ellipses For Random Output
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Ellipses For Random Output
~~~~~~~~~~~~~~~~~~~~~~~~~~
If the output of some operation in an example cannot be verified exactly,
an ellipsis (``...``) can be used in the expected output. This allows it
to match any substring in the actual output.

.. code ::

    Examples
    --------

    >>> desktop = Desktop("2021.1")
    pyaedt Info: pyaedt v...
    pyaedt Info: Python version ...

To allow this, ``doctest`` must be run with the option set to allow ellipses.

.. code ::

    pytest --doctest-modules -o ELLIPSIS file.py

> ``doctest`` Skip
> ~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer: ```suggestion
``doctest`` Skip
~~~~~~~~~~~~~~~~
```

The directive ``# doctest: +SKIP`` can be added to any line of a

> docstring example so that it is not executed. This is useful for examples

Reviewer: ```suggestion
docstring example so that it is not executed in `doctest-modules`. This is useful for examples
```

> that have really random output or that may cause problems with other tests

Reviewer: ```suggestion
that cannot run within ``pytest`` or have side-effects that will affect the other tests
```

if they are run during the ``doctest`` session.

> .. code ::

Reviewer: ```suggestion
.. code :: python
```
